### PR TITLE
Refactor AHA and split into sub-modules

### PR DESCRIPTION
**Pull request by abdel**
**Generic Components**
- DG is still being housed in `components` as it is not necessarily always tied to AHA and we could use it elsewhere
- PC Buffer, now known as `KNNBuffer`, now also lives in `components` as a more generalised version of a simple buffer with override/append options and a k-nearest neighbour lookup mechanism

**AHA Sub-Modules**
- `PatternMapper`. A separate module that maps arbitrary inputs to arbitrary outputs, used as basis for PM/PM_EC. It could probably be moved to components, given that it's pretty generic...?
- `PerforantPR`. A separate module containing all the build logic of PR, optimisation and pre/post processing.
- `PerforantHebb`. A separate sub-module containing the build logic for the local connections EC:CA3 and DG:CA3, and their optimisation logic and learning rule definitions.
- `MonosynapticPathway`. A separate sub-module containing the build logic for CA1 network and the CA3:CA1 connection network. It includes the logic for optimisation of two networks, and the test-time variation for recall where CA3:CA1 will drive the CA1 decoding.

**AHA Sub-Module Naming Changes**
- PS => renamed to DG
- PR => consolidated into `PerforantPR`
- EC_CA3, DG_CA3 => combined into `PerforantHebb`
- PC => renamed to CA3, initialised as `KNNBuffer`
- PM => naming unchanged, initialised from `PatternMapper`
- PM_EC => naming unchanged, initialised from `PatternMapper`
- CA1, CA3_CA1 => combined into `MonosynapticPathway`, `msp.py`

### CODE REVIEW

File: cls_module/cls_module/components/local_connection.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+"""LocalConnection class."""
+
+import torch.nn as nn
+
+class LocalConnection(nn.Linear):
```
Comment (affogato) [inline cls_module/cls_module/components/local_connection.py:5]: Can you please add some documentation for the class.
Include a description in the init method of the meaning of the parameters - mainly bias

File: cls_module/cls_module/components/local_connection.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  def __init__(self, input_features, output_features, bias=False):
+    super(LocalConnection, self).__init__(input_features, output_features, bias)
+
+    self.weight.requires_grad = False
+
+    if bias:
```
Comment (affogato) [inline cls_module/cls_module/components/local_connection.py:28]: if bias, wouldn't that mean you want grad to True?

File: cls_module/cls_module/memory/stm/aha/aha.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    Propagate
+    PR: BP Optimization step over
+  """
+
+  global_key = 'stm'
+  local_key = 'aha'
+
+  def is_hebbian_perforant(self):
+    k = 'hebbian_perforant'
+    if k in self.config and self.config[k]:
```
Comment (affogato) [inline cls_module/cls_module/memory/stm/aha/aha.py]: Would it be more concise to use this `dict.get(keyname, value)`
And then you don't need to define the additional variable `k`

File: cls_module/cls_module/memory/stm/aha/aha.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+  def build(self):
+    """Build AHA as short-term memory module."""
+
+    # Build the Dentae Gyrus
+    self.dg = DG(self.input_shape, self.config['dg']).to(self.device)
+    dg_output_shape = [1, self.config['dg']['num_units']]
+
+    # Build the Perforant Pathway
+    if self.is_hebbian_perforant():
```
Comment (affogato) [inline cls_module/cls_module/memory/stm/aha/aha.py:73]: If there is a conditional for which one to build, then can't you use the same instance variable name.
e.g. `self.perforant` 
More compact

File: cls_module/cls_module/memory/stm/aha/aha.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    # Compute DG Overlap
+    overlap = self.dg.compute_overlap(outputs['dg'])
+    losses['dg_overlap'] = overlap.sum()
+
+    # Perforant Pathway: Hebbian Learning
+    if self.is_hebbian_perforant():
+      ca3_cue, losses['dg_ca3'], losses['ec_ca3'], losses['ca3_cue'] = self.perforant_hebb(ec_inputs=inputs, dg_inputs=outputs['dg'])
+
+    # Perforant Pathway: Error-Driven Learning
+    if not self.is_hebbian_perforant():
```
Comment (affogato) [inline cls_module/cls_module/memory/stm/aha/aha.py]: why not use an `else` ?

File: cls_module/cls_module/memory/stm/aha/msp.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+  def reset(self):
+    if self.ca1_reset_params:
+      self.ca1.reset_parameters()
+
+    if self.ca3_ca1_reset_params:
+      self.ca3_ca1.reset_parameters()
+
+    # Reset the module optimizer
+    if self.ca1_reset_optim:
+      self.ca1_optimizer.state = defaultdict(dict)
```
Comment (affogato) [inline cls_module/cls_module/memory/stm/aha/msp.py:51]: i can't see where `dict` is defined .??

File: cls_module/cls_module/memory/stm/aha/msp.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        'output': encoding.detach()  # Designated output for linked modules
+    }
+
+    if self.training:
+      loss.backward()
+      self.ca1_optimizer.step()
+
+    return loss, outputs
+
+  def forward_ca3_ca1(self, inputs, targets):
```
Comment (affogato) [inline cls_module/cls_module/memory/stm/aha/msp.py:77]: I think it's good where it is at the moment.
Just be aware that this is technically considered part of the TSP

**Review by affogato — COMMENTED** [review]
not yet finished review

File: cls_module/cls_module/components/local_connection.py (hunk 2 above)
Comment (abdel) [reply cls_module/cls_module/components/local_connection.py:28]: Not necessarily.

You could want to bias the weighted sum with a constant value. I guess in this case it's always going to be default to zeros, but we could allow biasing with different constant values or random values/noise?

File: cls_module/cls_module/memory/stm/aha/aha.py (hunk 4 above)
Comment (abdel) [reply cls_module/cls_module/memory/stm/aha/aha.py:73]: Yep, agree.

File: cls_module/cls_module/memory/stm/aha/aha.py (hunk 5 above)
Comment (abdel) [reply cls_module/cls_module/memory/stm/aha/aha.py]: I don't recall anymore, but `else` makes sense. I think originally there was stuff between those two statements, but not necessarily anymore.

I will modify this.

File: cls_module/cls_module/memory/stm/aha/msp.py (hunk 7 above)
Comment (abdel) [reply cls_module/cls_module/memory/stm/aha/msp.py:77]: Yeah, I contemplated making them separate modules, but to make it as clean separation as possible, having this particular connection of the TSP as part of the MSP made the most sense. This is because they are both intertwined both during training, and especially during test time where it completely overrides CA1.

We can re-adjust later :)

File: cls_module/cls_module/memory/stm/aha/msp.py (hunk 6 above)
Comment (abdel) [reply cls_module/cls_module/memory/stm/aha/msp.py:51]: `dict` is a built-in Python keyword to define dicts. It is essentially setting `optimizer.state` to an empty dict i.e. `{}`.

I used this instead as by default PyTorch initialises states like this, so it made sense to just replicate that when resetting  in case something internal is expecting a `defaultdict` object for some reason.

File: cls_module/cls_module/components/local_connection.py (hunk 1 above)
Comment (abdel) [reply cls_module/cls_module/components/local_connection.py:5]: I will - but briefly, `LocalConnection` inherits from `nn.Linear` which is just a basic linear layer `z = w.x + b` where weights and bias are a typical matrix and vector respectively.

The difference in LocalConnection is that we a) override and control the initialisation of these parameters b) given that its local Hebbian learning, we turn off gradient learning in weight paramter (and in bias if used)

File: cls_module/cls_module/components/local_connection.py (hunk 2 above)
Comment (affogato) [reply cls_module/cls_module/components/local_connection.py:28]: To be honest, I don't totally understand :-)
I'm flagging it, and you change it if necessary. 
I'll get to understand this when I run it.

**Review by affogato — APPROVED** [review]
You should merge it when you've made any changes that you see fit based on my comments.
